Source/WebCore/platform/audio/cocoa/AudioEncoderCocoa.cpp

```diff
@@ -0,0 +1,385 @@
+    });
+}
+
+Ref<GenericPromise> AudioEncoderCocoa::flush()
+{
+    return invokeAsync(InternalAudioEncoderCocoa::queueSingleton(), [encoder = m_internalEncoder] {
+        return encoder->flush();
+    });
+}
+
+void AudioEncoderCocoa::reset()
+{
+    InternalAudioEncoderCocoa::queueSingleton().dispatch([encoder = m_internalEncoder] {
+        encoder->close();
+    });
+}
+
+void AudioEncoderCocoa::close()
+{
+    InternalAudioEncoderCocoa::queueSingleton().dispatch([encoder = m_internalEncoder] {
+        encoder->close();
+    });
+}
+
+InternalAudioEncoderCocoa::InternalAudioEncoderCocoa(const Config& config, AudioEncoder::DescriptionCallback&& descriptionCallback, AudioEncoder::OutputCallback&& outputCallback)
+    : m_descriptionCallback(WTFMove(descriptionCallback))
+    , m_outputCallback(WTFMove(outputCallback))
+    , m_config(config)
+{
+}
+
+String InternalAudioEncoderCocoa::initialize(const String& codecName)
+{
+    auto result = AudioDecoderCocoa::isCodecSupported(codecName);
+    if (!result)
+        return result.error();
+
+    std::tie(m_codec, m_pcmFormat) = *result;
+    if (m_codec == 'vorb')
+        return "Vorbis encoding is not supported"_s;
+    if (m_codec == kAudioFormatFLAC)
+        return "FLAC encoding is not supported"_s;
+    if (m_codec == kAudioFormatMPEGLayer3)
+        return "MP3 encoding is not supported"_s;
+
+    if (m_codec == kAudioFormatOpus) {
+        if (m_config.numberOfChannels > 2)
+            return "Opus with more than two channels is not supported"_s;
+
+        if (m_config.opusConfig && m_config.opusConfig->isOggBitStream)
+            return "Opus Ogg format is unsupported"_s;
+
+        auto config = m_config.opusConfig.value_or(AudioEncoder::OpusConfig { });
+        m_frameDuration = MediaTime(config.frameDuration, 1000000);
+        m_complexity = config.complexity;
+        m_packetlossperc = config.packetlossperc;
+        m_useinbandfec = config.useinbandfec;
+    } else if ((m_codec == kAudioFormatMPEG4AAC || m_codec == kAudioFormatMPEG4AAC_HE || m_codec == kAudioFormatMPEG4AAC_LD || m_codec == kAudioFormatMPEG4AAC_HE_V2 || m_codec == kAudioFormatMPEG4AAC_ELD)) {
+        m_isADTS = m_config.isAacADTS.value_or(false);
+        if (m_isADTS)
+            return "AAC ADTS format is unsupported"_s;
+    }
+
+    if (m_pcmFormat)
+        m_outputDescription = CAAudioStreamDescription { double(m_config.sampleRate), uint32_t(m_config.numberOfChannels), *m_pcmFormat, CAAudioStreamDescription::IsInterleaved::Yes };
+    else {
+        AudioStreamBasicDescription asbd { };
+        asbd.mFormatID = m_codec;
+        asbd.mSampleRate = double(m_config.sampleRate);
+        asbd.mChannelsPerFrame = uint32_t(m_config.numberOfChannels);
+        if (m_frameDuration)
+            asbd.mFramesPerPacket = m_config.sampleRate * m_frameDuration->toDouble();
+        m_outputDescription = asbd;
+    }
+    m_codecString = codecName.isolatedCopy();
```

Comment:
Do we really need this, or could we use an enumeration instead?
That could prevent issues with Strings...

Reply:
I'm happy to change this as a follow-up, right now that's pre-existing and how the GStreamer implementation uses it. Don't want to touch it here.

Reply:
we probably don't need the isolatedCopy but we generate the `activeConfiguration` on the encode workqueue. So the strings needs to work across threads.